test/spec/features/service/exec_spec.rb

```diff
@@ -36,22 +36,83 @@
     expect(k.code).to eq(126)
   end
 
-  it 'runs a command inside a service with tty' do
-    k = kommando("kontena service exec -it test-1 sh")
+  describe '--tty' do
+    it 'runs a command inside a service' do
+      k = kommando("kontena service exec -it test-1 sh")
+
+      k.out.on("#") do
+        k.in << "ls -la /\r"
+        k.out.on "lib64" do
+          k.in << "exit\r"
+        end
+      end
+      expect(k.run).to be_truthy
+    end
+
+    it 'runs a command with tty control input' do
+      k = kommando("kontena service exec -it test-1 sh")
 
-    k.out.on("#") do
-      k.in << "ls -la /\r"
-      k.out.on "lib64" do
-        k.in << "exit\r"
+      k.out.on("#") do
+        k.in << "sleep 10 && echo ok\r"
+        sleep 0.1
+        k.in << "\x03" # Ctrl-C -> SIGINT
+        sleep 0.1
+        k.in << "\x04" # Ctrl-D -> EOF
       end
+
+      expect(k.run).to be_truthy
+      expect(k.out).to match /\^C/
+    end
+
+    it 'runs a command with non-ascii input' do
+      k = kommando("kontena service exec -it test-1 sh")
+
+      k.out.on("#") do
+        k.in << "echo f\u00e5\u00e5 | LANG=C.UTF-8 rev\r"
+        k.out.on("#") do
+          k.in << "exit\r"
+        end
+      end
+
+      expect(k.run).to be_truthy
+      expect(k.out).to match /\u00e5\u00e5f/
+    end
+
+    # RuntimeError : stdin read JSON::GeneratorError: partial character in source, but hit end
+    pending 'runs a command with chunked UTF-8 input' do
+      k = kommando("kontena service exec -it test-1 sh")
+
+      k.out.on("#") do
+        data = "echo f\u00e5\u00e5\r"
+
+        # write slowly, one byte at a time
+        data.encode('UTF-8').force_encoding('ASCII-8BIT').each_char do |c|
+          k.in << c
+          sleep 0.05
```

Comment:
This is more realistic: if some of the UTF-8 input bytes are slower to arrive, then the CLI will fail with partial character errors.